Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift

```diff
@@ -0,0 +1,94 @@
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+
+#if ENABLE_SWIFTUI && compiler(>=6.0)
+
+import Foundation
+internal import WebKit_Internal
+
+@_spi(Private)
+public struct URLScheme_v0: Sendable, Hashable, Equatable {
+    let rawValue: String
+
+    public init?(_ rawValue: String) {
+        guard let canonicalScheme = WKWebViewConfiguration._canonicalScheme(forURLScheme: rawValue) else {
+            return nil
+        }
+
+        self.rawValue = canonicalScheme
+    }
+}
+
+@_spi(Private)
+public enum URLSchemeTaskResult_v0: Sendable {
+    case response(URLResponse)
+
+    case data(Data)
+}
+
+@_spi(Private)
+public protocol URLSchemeHandler_v0 {
+    associatedtype TaskSequence: AsyncSequence<URLSchemeTaskResult_v0, any Error>
+
+    func reply(for request: URLRequest) -> TaskSequence
+}
+
+// MARK: Adapters
+
+final class WKURLSchemeHandlerAdapter: NSObject, WKURLSchemeHandler {
+    private let wrapped: any URLSchemeHandler_v0
+
+    private var tasks: [ObjectIdentifier: Task<Void, Never>] = [:]
+
+    init(wrapping wrapped: any URLSchemeHandler_v0) {
+        self.wrapped = wrapped
+    }
+
+    func webView(_ webView: WKWebView, start urlSchemeTask: any WKURLSchemeTask) {
+        let task = Task {
+            do {
+                for try await result in wrapped.reply(for: urlSchemeTask.request) {
+                    switch result {
+                    case .response(let response):
+                        urlSchemeTask.didReceive(response)
+
+                    case .data(let data):
+                        urlSchemeTask.didReceive(data)
+                    }
+                }
+
+                urlSchemeTask.didFinish()
+            } catch {
+                urlSchemeTask.didFailWithError(error)
+            }
+        }
+
+        tasks[ObjectIdentifier(urlSchemeTask)] = task
+    }
+
+    func webView(_ webView: WKWebView, stop urlSchemeTask: any WKURLSchemeTask) {
+        tasks.removeValue(forKey: ObjectIdentifier(urlSchemeTask))?.cancel()
```

Comment:
Something to double check is how the cancellation propagation works here. I'm not sure if it's partially in the client's control, and if so whether we can do things to make it more robust to prevent misuse (ex: checking `Task.isCancelled` on their behalf before processing results maybe?). Don't make changes for this right now though.